[dynamo] Relax guard introduced when tracing `__call__` on user defined object

StrongerXi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #152395
* #152369

This relaxes the guard introduced in #100444 (which aggressively guard
on the object id, despite Dynamo is just tracing its `__call__` method.

This allows users to bypass the high compilation time issue in #150706
by compiling transformer blocks only. Without this patch, we'd get lots
of unnecessary recompilation, as the block has difference attention
processor instances.

Compiling blocks only _significantly_ speeds up compilation process
(from ~310s to ~32s), and even speeds up e2e performance for some reason
(7.83s to 7.67s).

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

StrongerXi committed:
[dynamo] Relax guard introduced when tracing `__call__` on user defined object

This relaxes the guard introduced in #100444 (which aggressively guard
on the object id, despite Dynamo is just tracing its `__call__` method.

This allows users to bypass the high compilation time issue in #150706
by compiling transformer blocks only. Without this patch, we'd get lots
of unnecessary recompilation, as the block has difference attention
processor instances.

Compiling blocks only _significantly_ speeds up compilation process
(from ~310s to ~32s), and even speeds up e2e performance for some reason
(7.83s to 7.67s).

[ghstack-poisoned]

anijain2305 commented on torch/_dynamo/variables/user_defined.py:
```diff
@@ -888,6 +888,8 @@ def call_method(
                 if method is torch.nn.Module.__init__:
                     method = unpatched_nn_module_init
+                if source:
+                    install_guard(source.make_guard(GuardBuilder.FUNCTION_MATCH))
```
I am concerned that this will add a lot of new guards (even though this might be the right thing to do). Maybe skip this one in this PR, and then follow up with another PR with some more benchmarking?

StrongerXi replied:
Lemme localize the fix to that one site instead, that should cover us in practice without introducing too many new guards.

StrongerXi committed:
Update on "[dynamo] Relax guard introduced when tracing `__call__` on user defined object"


This relaxes the guard introduced in #100444 (which aggressively guard
on the object id, despite Dynamo is just tracing its `__call__` method.

This allows users to bypass the high compilation time issue in #150706
by compiling transformer blocks only. Without this patch, we'd get lots
of unnecessary recompilation, as the block has difference attention
processor instances.

Compiling blocks only _significantly_ speeds up compilation process
(from ~310s to ~32s), and even speeds up e2e performance for some reason
(7.83s to 7.67s).

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

StrongerXi commented:
@pytorchbot merge

StrongerXi commented:
@pytorchbot merge -f "unrelated failure"
